R/get_stats.R

```diff
@@ -133,35 +133,63 @@ getRhythmStats = function(
 #' @seealso [getRhythmStats()], [getStatsIntervals()]
 #'
 #' @export
-getDiffRhythmStats = function(fit, rhyStats, condLevels) {
+getDiffRhythmStats = function(fit, rhyStats, condLevels = NULL) {
   cond = .SD = peak_phase = trough_phase = NULL
 
   assertClass(fit, 'limorhyde2')
   assertTRUE(fit$nConds >= 2)
   assertDataTable(rhyStats)
   assertTRUE(attr(rhyStats, 'statType') == 'rhy')
   assertTRUE('cond' %in% colnames(rhyStats))
-  assertAtomicVector(condLevels, len = 2L)
   assertSubset(condLevels, fit$condLevels)
   assertSubset(condLevels, levels(rhyStats$cond))
 
+  if (is.null(condLevels)) condLevels = unique(rhyStats$cond)
   d0 = rhyStats[cond %in% condLevels]
   set(d0, j = 'cond', value = factor(d0$cond, condLevels))
   data.table::setorderv(d0, 'cond')
+  d0[, cond2Int := as.integer(cond)]
 
   fitType = attr(rhyStats, 'fitType')
   byCols = c('feature', if (fitType == 'posterior_samples') 'posterior_sample')
   cols = c('mean_value', 'peak_trough_amp', 'rms_amp', 'peak_phase', 'trough_phase')
 
-  diffRhyStats = d0[, lapply(.SD, diff), by = byCols, .SDcols = cols]
+  diffRhyStats = d0[,
+        .SD[.SD, on = .(cond2Int < cond2Int),
+          .(cond_1 = cond, cond_2 = i.cond,
+            mean_value, peak_trough_amp, rms_amp, peak_phase, trough_phase,
+            i.mean_value, i.peak_trough_amp, i.rms_amp, i.peak_phase,
+            i.trough_phase)],
+    by = feature]
+  diffRhyStats = diffRhyStats[,
+        .(feature, cond_1, cond_2,
+        mean_value = mean_value - i.mean_value,
+        peak_trough_amp = peak_trough_amp - i.peak_trough_amp,
+        rms_amp = rms_amp - i.rms_amp,
+        peak_phase = peak_phase - i.peak_phase,
+        trough_phase = trough_phase - i.trough_phase)]
+  diffRhyStats = diffRhyStats[!is.na(mean_value)]
+
   diffRhyStats[, peak_phase := centerCircDiff(peak_phase, fit$period)]
   diffRhyStats[, trough_phase := centerCircDiff(trough_phase, fit$period)]
   data.table::setnames(diffRhyStats, cols, paste0('diff_', cols))
 
+  #create data.table of conditions
+  pairDt = getCondPairs(condLevels)
+  pairIter = iterators::iter(pairDt, by = 'row')
+
   # calculate rms difference in rhythmic fit between conditions
   featureIdx = rownames(fit$coefficients) %in% unique(rhyStats$feature)
-  rmsDiffRhy = getRmsDiffRhy(fit, condLevels, fitType, featureIdx)
-  diffRhyStats = merge(diffRhyStats, rmsDiffRhy, sort = FALSE)
+  rmsDiffRhy = foreach(cond_1 = pairDt[, cond_1], cond_2 = pairDt[, cond_2],
```

Comment:
Can you iterate over rows without setting up an `iter`?

Reply:
No, I don't think so. I want to solve the issue with `iter`, because I've used it multiple times with no issue.
 I think the problem is something in `getCondPairs`.

Reply:
I haven't been able to find a good way to get all of the pairs with just data.table, and I wanted them to be factors so that the "baseline" value would come first.

Reply:
If `iter` isn't working, then I would just do `foreach(i = 1:nrow(d) ...`, then inside the loop use `condPairs[i]$...`

Theoretically not as efficient and not as readable, but `condPairs` is so small the difference will be negligible.